WalletWasabi/Tor/Control/TorControlClient.cs

```diff
@@ -90,31 +134,115 @@ public Task<TorControlReply> ResetOwningControllerProcessConfAsync(CancellationT
 		/// <param name="command">A Tor control command which must end with <c>\r\n</c>.</param>
 		public async Task<TorControlReply> SendCommandAsync(string command, CancellationToken cancellationToken = default)
 		{
+			using var _ = await MessageLock.LockAsync(cancellationToken).ConfigureAwait(false);
+
 			Logger.LogTrace($"Client: About to send command: '{command.TrimEnd()}'");
 			await PipeWriter.WriteAsync(new ReadOnlyMemory<byte>(Encoding.ASCII.GetBytes(command)), cancellationToken).ConfigureAwait(false);
 			await PipeWriter.FlushAsync(cancellationToken).ConfigureAwait(false);
 
-			TorControlReply reply = await TorControlReplyReader.ReadReplyAsync(PipeReader, cancellationToken).ConfigureAwait(false);
-
+			TorControlReply reply = await SyncChannel.Reader.ReadAsync(cancellationToken).ConfigureAwait(false);
 			return reply;
 		}
 
-		protected virtual void Dispose(bool disposing)
+		/// <summary>Allows the caller to read Tor events using <c>await foreach</c>.</summary>
+		/// <remarks>Processing of replies does not block other readers.</remarks>
+		/// <example>
+		/// <code>
+		/// await foreach (TorControlReply reply in TorControlClient.ReadEventsAsync())
+		/// {
+		///    Console.WriteLine(reply.ToString());
+		/// }
+		/// </code> 
+		/// </example>
+		public async IAsyncEnumerable<TorControlReply> ReadEventsAsync([EnumeratorCancellation] CancellationToken cancellationToken = default)
 		{
-			if (!_disposedValue)
+			Channel<TorControlReply> channel = Channel.CreateUnbounded<TorControlReply>(Options);
+			List<Channel<TorControlReply>> newList;
+
+			try
+			{
+				lock (AsyncChannelsLock)
+				{
+					newList = new(AsyncChannels);
+					newList.Add(channel);
+
+					AsyncChannels = newList;
+				}
```

Comment:
Why do we need the `newList`? `AsyncChannels` are modified inside a lock, so there are no thread-safety concerns.

Reply:
There are several reasons:

1. You can't modify a list if it is being iterated (i.e. `foreach (Channel<TorControlReply> channel in AsyncChannels)` would throw a collection-modified-exception, I'm not sure about the exact exception name now.)
2. `AsyncChannels` are modified inside a lock but *iteration* of the list is NOT. To make the code thread-safe you would need to wrap iteration of the `AsyncChannels` in `lock () { ... }` too. I did not want to do that because `lock () {...}` cannot contain `async` commands and async lock is much heavier than `lock`. So there is a trick for these situations that you use a copy-on-write list technique and that technique simply creates a new list once the original list needs to be modified so you have no issues with concurrent modification in a `foreach` and you only swap the old list with the new one under `lock` so you don't have any concurrency issues.

Not sure if it is explained well. But I tried. :)

Reply:
You explained it perfectly well, I got it now. Thank you.